### PR TITLE
Completes 10. GET/api/articles-queries

### PR DESCRIPTION
**Pull request by Tayamul**
git add __tests__/app.test.js controllers/articles.controller.js models/articles.model.js

### CODE REVIEW

File: __tests__/app.test.js [hunk 1]
```diff
@@ -63,20 +63,124 @@ describe("GET/api/articles", () => {
         });
       });
   });
-  test("200: articles should be sorted by date in descending order by default ", () => {
+  test("200: should accept a topic query which filters the articles by the topic value specified", () => {
+    return request(app)
+      .get("/api/articles?topic=mitch")
+      .expect(200)
+      .then(({ body: { articles } }) => {
+        expect(articles).toBeInstanceOf(Array);
```
Comment (B7Bprog) [inline __tests__/app.test.js:71]: Please check the length of the articles array.

File: __tests__/app.test.js [hunk 2]
```diff
@@ -63,20 +63,124 @@ describe("GET/api/articles", () => {
...
+      .then(({ body: { articles } }) => {
+        expect(articles).toEqual([]);
+      });
+  });
+  test("200: should respond with all the articles if topic query is not provided", () => {
+    return request(app)
+      .get("/api/articles")
+      .expect(200)
+      .then(({ body: { articles } }) => {
+        expect(articles).toBeInstanceOf(Array);
```
Comment (B7Bprog) [inline __tests__/app.test.js:93]: Array length check missing.

File: __tests__/app.test.js [hunk 3]
```diff
@@ -63,20 +63,124 @@ describe("GET/api/articles", () => {
...
+        expect(msg).toBe("Bad Request");
+      });
+  });
+  test("200: responds with an array of articles given all three valid queries", () => {
+    return request(app)
+      .get("/api/articles?topic=mitch&sort_by=author&order=asc")
+      .expect(200)
+      .then(({ body: { articles } }) => {
+        expect(articles).toBeInstanceOf(Array);
+        expect(articles).toBeSortedBy("author");
```
Comment (B7Bprog) [inline __tests__/app.test.js:173]: Checking array length missing.

File: __tests__/app.test.js [hunk 4]
```diff
@@ -63,20 +63,124 @@ describe("GET/api/articles", () => {
...
+              author: expect.any(String),
+              title: expect.any(String),
+              article_id: expect.any(Number),
+              topic: "mitch",
+              created_at: expect.any(String),
+              votes: expect.any(Number),
+              comment_count: expect.any(String),
+            })
+          );
+        });
```
Comment (B7Bprog) [inline __tests__/app.test.js:186]: Please add another test where the topic doesn't exist, e. g. `?topic=bananas`. Expecting a status 404, 'not found'.

**Review by B7Bprog — CHANGES_REQUESTED** [review]
Hi Tayamul, nice work so far, please see my comments, refactor / add code and send us another pr with it.

File: __tests__/app.test.js (hunk 4 above)
Comment (Tayamul) [reply __tests__/app.test.js:186]: I have already tested for this situation which returns an empty array and a status 200 if the topic does not exist. 
I was thinking of creating a variable validTopicQueries = ['name of all the topics inside this array'], and then having an if statement where if the values inside validTopicQueries does not include in the topic, promise reject it with a 404.
But I had a conflicting thought about it where the topics in out testData and in our devData has different topics. So the test would not pass in different environments. But am I wrong to think in that way? And just carry on with the above solution for a 404 error?

**Review by robert112233 — APPROVED** [review]
Good work, merge 👍

**Pull request merged**